pkgs/by-name/li/libfprint-2-tod1-broadcom/package.nix

```diff
@@ -0,0 +1,81 @@
+{
+  autoPatchelfHook,
+  fetchzip,
+  lib,
+  libfprint-tod,
+  openssl,
+  patchelfUnstable,
+  stdenv,
+}:
+
+# Based on ideas from (using a wrapper library to redirect fopen() calls to firmware files):
+#   * https://tapesoftware.net/replace-symbol/
+#   * https://github.com/NixOS/nixpkgs/pull/260715
+let
+  version = "5.12.018";
+  src = fetchzip {
+    url = "http://dell.archive.canonical.com/updates/pool/public/libf/libfprint-2-tod1-broadcom/libfprint-2-tod1-broadcom_${version}.orig.tar.gz";
+    hash = "sha256-0C2PpYpEJNrU+8NT95w4QV0J5nHQisMY94Czw3jQOzw=";
+  };
+
+  unpacked = stdenv.mkDerivation {
+    pname = "libfprint-2-tod1-broadcom-unpacked";
+    inherit src version;
+
+    dontBuild = true;
+
+    installPhase = ''
+      mkdir $out
+      cp -r . $out
+    '';
+  };
+
+  # wraps `fopen()` for finding firmware files
+  wrapperLibName = "wrapper-lib.so";
+  wrapperLib = stdenv.mkDerivation {
```

Comment:
It probably would; to be honest I didn't know about it before and didn't test it.

Still, this would affect the whole process, not only the calls in the library `libfprint-2-tod-1-broadcom.so` into which I inject the wrapper library, would it not? I guess in principle that is okay and should not be a problem in this case. But one could see it as a benefit that the wrapper library affects only those calls that are really necessary, and in particular only those from the corresponding library.

Personally I would probably still opt for this "less invasive" approach (if I understood `libdirect` right and my approach does in fact affect only necessary calls). But I am not too experienced in this respect, so if you think the added complexity of the wrapper-lib is a major drawback, I can try an alternative implementation via `libredirect`.

Reply:
I do not really recall how ELF symbol resolution and `--rename-dynamic-symbols` specifically works. You are probably right that this is less invasive than `LD_PRELOAD`.

Ideally, we would modify `libredirect` to allow using it with `--rename-dynamic-symbols` but we do not need to do that here.

Reply:
Just for future reference: The technique on which `--rename-dynamic-symbols` is based on is described [here](https://tapesoftware.net/replace-symbol/) (also referenced in my code). The rough idea is explained under the heading **patchelf --replace-symbol**. Also, I just realized that the page mentions that this approach is less invasive than `LD_PRELOAD` (under the heading **Advantages**).

So, for now I'll leave it with the current implementation via the wrapper-library and mark this conversation as resolved.